Feature/add enzyme testing infrastructure

Akin909 opened the pull request:
Initial infrastructure setup for using `jest` and `enzyme` to begin testing react components,
this PR includes:

* configuration of `jest` in package.json
* setting up a jests setup file
* adding separate commands to test `react` components (as this uses jest not mocha)
* create separate `/components/tests/` dir so these test remain in proximity to their subjects for easier importing and also to avoid collision with mocha (not an issue I've observed just one I worry about)
* Add `redux-mock-store` to facilitate testing redux connected components.

As discussed in #1128 once this PR goes in I can begin to look at gradually add tests hopefully most/all/hopefully not alone react components

NB: The tests I've included were more to test that enzyme and redux mock store and the commands were all working correctly, will be changing those subsequently

Akin909 commented:
Update: a little blocked due to conflicting typings with `mocha/jest` both using `it` and `describe` syntax with clashing typedefs 😦

bryphe commented:
> Update: a little blocked due to conflicting typings with mocha/jest both using it and describe syntax with clashing typedefs 😦

Ouch, is there a way to exclude particular typedefs? http://www.typescriptlang.org/docs/handbook/tsconfig-json.html

Like, for the `jest` tests, can we specifically ignore the mocha definitions? And vice-versa?

One other question - is `jest` necessary for integrating `enzyme`? Or could we start by integrating `jest` to our `mocha` infrastructure, and then look at switching to `jest` as part 2?

Akin909 commented:
@bryphe mocha can run with enzyme, the reason I started with `jest` which I definitely appreciate is the bigger move is because the process of trying to get the test to run with mocha was proving very time consuming and not productive, installing jest just led to everything working plus within half an hour also tbh its the tool I know how to use/ setup to work with `enzyme`.

The only real issue is with typings the actual test for both `mocha` and `jest` work fine independent of each other they just share the same names, ill definitely look into setting up the tsconfig to ignore certain types

bryphe commented:
> @bryphe mocha can run with enzyme, the reason I started with jest which I definitely appreciate is the bigger move is because the process of trying to get the test to run with mocha was proving very time consuming and not productive, installing jest just led to everything working plus within half an hour also tbh its the tool I know how to use/ setup to work with enzyme.

👍 - makes sense! Thanks for the details. Not worth wasting time if the integration is much better with `jest`!

Might be worth looking at the `@types, typeRoots, and types` section here:
http://www.typescriptlang.org/docs/handbook/tsconfig-json.html. Perhaps we can explicitly specify `types` for these new tests?

Akin909 commented:
Made some progress here splitting tests out into `ui-test` in root folder and removing automatic global types from `tsconfigs` then specifying desired globals in the certain places so for browser `types: ["mocha"]` only global type that is used I believe and in `ui-test`, `types: ["jest"]`

Seems to resolve the duplicate but now need to resolve issue with dependencies in the react components, might make the test simpler and do that in another PR

Akin909 commented:
@bryphe finally got this working 🍾 a few things to note:

* To get round duplicate type issues I turned off automatic type inclusion by setting `types: [mocha]`, in **all** the tsconfig's this has to be in all tsconfigs as if one is missed it will automatically include global jest and mocha types which conflict. This doesnt impact other oni functionality as those are the only cases where we depend on those.

* I had a lot of *random* linting errors come up all of a sudden hence all the import statement shuffling not sure why the linter suddenly noticed a bunch of issues.

* I added mocks for a lot of functionality that oni components *at least the ones I tested* depended on.

* Added a section to the package json for `jest` possibly this can be moved into a config file however **whenever** I try it doesnt work, I'm definitely missing something not sure what.

Akin909 commented:
I wonder if we have directories to exclude from coverage as `snapshots` and `mocks` etc. might bring down coverage if they arent excluded

bryphe commented:
> @bryphe finally got this working 🍾 a few things to note:

Awesome, nice work getting this going!

> I wonder if we have directories to exclude from coverage as snapshots and mocks etc. might bring down coverage if they arent excluded

Is `jest` hooked up to code coverage? If not, that would be a good next step for us - looking at how to integrate `jest` with `codecov`. If we had that, it would get incorporated in the reports - and we'd get coverage data for our React components, too!

> I added mocks for a lot of functionality that oni components at least the ones I tested depended on.

Some of the mocks look like they might be duplicated from https://github.com/onivim/oni/blob/master/browser/test/Mocks/index.ts . As a next step, would be good to see if we could put these Mocks in a common place such that they are accessible for both the mocha tests and jest tests - maintaining two different sets of Mocks will be painful.

Akin909 commented on package.json:
```diff
@@ -252,13 +287,17 @@
         "electron-rebuild": "1.6.0",
+        "enzyme": "^3.3.0",
+        "enzyme-adapter-react-16": "^1.1.1",
+        "enzyme-to-json": "^3.3.1",
```
👍 😄 no worries I'm looking forward to the added safety net of tested react components

Akin909 commented:
@bryphe managed to remove settings from the `package.json` as that becoming quite complex,

Re. coverage jest has loads of options for generating coverage and outputting it etc. never personally configured it myself but think it should be *hopefully* quite straight forward can look into after this.

Re mocks I think it'd be best to keep them separate because jest mocks tend to take the form of `thingImMocking: jest.fn() or jest.mockImplementation` etc. I think if in a common place or even shared the chances of things imploding due to duplicate type detection is higher and also because jest mocks aren't really generalisable they tend be aimed at the library and its test mocha couldn't call a module mocked for jest and successfully use it. Thankfully mocking in jest is actually really flexible and straight forward usually you simply need to call `jest.mock(package | filename)`

The reason I've gone so far as to create mocks for those files is given how they fit into oni it seemed we would eventually want to have slightly more nuance mocking for those

bryphe commented:
> @bryphe managed to remove settings from the package.json as that's becoming quite complex,

Awesome, thanks!

> Re mocks I think it'd be best to keep them separate because jest mocks tend to take the form of thingImMocking: jest.fn() or jest.mockImplementation etc. I think if in a common place or even shared the chances of things imploding due to duplicate type detection is higher and also because jest mocks aren't really generalisable they tend be aimed at the library and its tests.

Ah ok, I see. That seems reasonable. The UI components should have a more streamlined set of dependencies, anyway. We can start with that and see where it goes 👍 

Thanks again for putting this together!

Akin909 commented:
@bryphe don't think I can merge this since it seems the macOS tests are failing all round (the build seems to be timing out at some point) and if I re-push ill loose the review approval 😆

bryphe commented:
Ah ya, the OSX tests are in rough shape right now, unfortunately! I'm working on stabilizing them... Will bring this in now 👍 Thanks again for working on this, @Akin909 - nice job!
